File: typescript/bc-lifehash/src/color-func.ts
import { Color, modulo } from './color.js';

export type ColorFunc = (t: number) => Color;

export function reverse(c: ColorFunc): ColorFunc {
    return (t: number) => c(1.0 - t);
}

export function blend2(color1: Color, color2: Color): ColorFunc {
    return (t: number) => color1.lerpTo(color2, t);
}

export function blend(colors: readonly Color[]): ColorFunc {
    const count = colors.length;
    if (count === 0) return blend2(Color.BLACK, Color.BLACK);
    if (count === 1) return blend2(colors[0], colors[0]);
    if (count === 2) return blend2(colors[0], colors[1]);
    const captured = colors.slice();
    return (t: number) => {
        if (t >= 1.0) return captured[count - 1];
        if (t <= 0.0) return captured[0];
        const segments = count - 1;
        const s = t * segments;
        const segment = Math.trunc(s);
        const segmentFrac = modulo(s, 1.0);
        const c1 = captured[segment];
        const c2 = captured[segment + 1];
        return c1.lerpTo(c2, segmentFrac);
    };
}

File: swift/BCLifeHash/Sources/BCLifeHash/ColorFunc.swift
typealias ColorFunc = (Double) -> Color

func reverse(_ c: @escaping ColorFunc) -> ColorFunc {
    { t in c(1.0 - t) }
}

func blend2(_ color1: Color, _ color2: Color) -> ColorFunc {
    { t in color1.lerpTo(color2, t) }
}

func blend(_ colors: [Color]) -> ColorFunc {
    let count = colors.count
    switch count {
    case 0:
        return blend2(.black, .black)
    case 1:
        return blend2(colors[0], colors[0])
    case 2:
        return blend2(colors[0], colors[1])
    default:
        return { t in
            if t >= 1.0 {
                return colors[count - 1]
            }
            if t <= 0.0 {
                return colors[0]
            }
            let segments = count - 1
            let s = t * Double(segments)
            let segment = Int(s)
            let segmentFrac = modulo(s, 1.0)
            let c1 = colors[segment]
            let c2 = colors[segment + 1]
            return c1.lerpTo(c2, segmentFrac)
        }
    }
}
